src/metabase/upload.clj

```diff
@@ -170,19 +179,38 @@
     "unnamed_column"
     (u/slugify (str/trim raw-name))))
 
+(defn- is-pk?
+  [[col-name type]]
+  (and (#{"id" "pk"} (u/lower-case-en (name col-name)))
+       (isa? type ::int)))
+
+(defn- ->ordered-maps-with-pk-column
+  "Sets appropriate type information on the first PK column it finds, otherwise adds a new one.
+
+  Returns a *map* with two keys: `:extant-columns` and `:generated-columns`."
+  [name-type-pairs]
+  (if-let [pk-name (first (keep (fn [[name _type :as p]]
+                                  (when (is-pk? p) name))
+                                name-type-pairs))]
+    {:extant-columns    (assoc (ordered-map/ordered-map name-type-pairs) pk-name ::int-pk)
+     :generated-columns (ordered-map/ordered-map)}
+    {:extant-columns    (ordered-map/ordered-map name-type-pairs)
+     :generated-columns (ordered-map/ordered-map :id ::auto-incrementing-int-pk)}))
```

Comment:
it's a little awkward juggling these guys around, but I thought it was less clumsy than the `{:type ::pk :opts foo :exclude-from-insert? true}` way I tried previously. At least it keeps the actual column-specs isomorphic.